Rsite/source/tutorial/CustomLossGluon.md

```diff
@@ -0,0 +1,235 @@
+loss = ContrastiveLoss(margin=6.0)
+```
+
+Start the training loop:
+
+
+```python
+for epoch in range(10):
+    
+    for i, data in enumerate(train_dataloader):
+        
+        image1, image2, label = data
+        with autograd.record():
+            output1, output2 = model(image1, image2)
+            loss_contrastive = loss(output1, output2, label)
+        loss_contrastive.backward()
+        trainer.step(image1.shape[0])
+        
+        print("Epoch number {}\n Current loss {}\n".format(epoch, loss_contrastive.mean().asscalar()))
+```
+
+### Test the trained Siamese network
+During inference we compute the Euclidean distance between the output vectors of the Siamese network. High distance indicates dissimilarity, low values indicate similarity.  
+
+
+```python
+for i, data in enumerate(test_dataloader):
+
+    img1, img2, label = data
+    output1, output2 = model(img1, img2)
+
+    print("Euclidean Distance:", mx.ndarray.sqrt(mx.ndarray.sum(mx.ndarray.square(output1 - output2))).asscalar(), "Test label", label[0].asscalar())
+    
+    fig, (ax0, ax1) = plt.subplots(ncols=2, figsize=(10, 5)) 
+    ax0.imshow(img1.asnumpy()[0,0,:,:], cmap='gray')
+    ax0.axis('off')
+    ax1.imshow(img2.asnumpy()[0,0,:,:], cmap='gray')
+    ax1.axis("off")
+    plt.show()
+
+    break
+```
+
+    Euclidean Distance: 2.494767 Test label 1.0
+
+
+
+![png](CustomLossGluon2.png)
+
+
+### Common Pitfalls with custom loss functions
+
+When customizing loss functions, we may encounter certain pitfalls. If the loss is not decreasing as expected or if forward/backward pass is crashing, then one should check the following:
+
+#### Activation function in the last layer:
+Verify whether the last network layer uses the correct activation function: for instance in binary classification tasks we need to apply a sigmoid on the output data. If we use this activation in the last layer and define a loss function like Gluon's SigmoidBinaryCrossEntropy, we would basically apply sigmoid twice and the loss would not converge as expected. If we don't define any activation function, Gluon will per default apply ```relu```.
+
+####  Intermediate loss values:
+In our example, we computed the square root of squared distances between 2 images: ```F.sqrt(distances_squared)```. If images are very similar we take the sqare root of a value close to 0, which can lead to *NaN* values. Adding a small epsilon to ```distances_squared``` avoids this problem.
+
+#### Shape of intermediate loss vectors:
+Computations like ```F.sum``` collapse dimensions and this could have side effects. In most cases having the wrong shape will lead to an error, as soon as we compare data with labels. But in some cases, we may be able to normally run the training, but it does not converge. For instance, if we don't set ```keepdims=True``` in our customized loss function, the example will run fine but not converge. 
+
+#### Differentiable
```

Comment:
Since this is long, a summary paragraph or conclusion seems like it would tie things up nicely